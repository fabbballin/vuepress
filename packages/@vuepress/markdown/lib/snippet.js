const { fs } = require('@vuepress/shared-utils')

module.exports = function snippet (md, options = {}) {
  const root = options.root || process.cwd()

  function parser (state, startLine, endLine, silent) {
    const pos = state.bMarks[startLine] + state.tShift[startLine]
    const max = state.eMarks[startLine]

    // if it's indented more than 3 spaces,
    // it should be a code block
    if (state.sCount[startLine] - state.blkIndent >= 4) {
      return false
    }

    if (!/^<<</.test(state.src)) {
      return false
    }

    if (silent) {
      return true
    }

    const start = pos + 3
    const end = state.skipSpacesBack(max, pos)
    const rawPath = state.src.slice(start, end).trim().replace(/^@/, root)
    const filename = rawPath.split(/[{\s]/).shift()
    const content = fs.existsSync(filename) ? fs.readFileSync(filename).toString() : 'Not found: ' + filename
    const meta = rawPath.replace(filename, '')

    const langExtWithMeta = () => {
      if (meta && /lang/.test(meta)) {
        const langSet = meta.match(/lang=\w+/)[0]
        if (langSet) {
          const lang = langSet.split('=')[1]
          const extractedMeta = meta.replace(langSet, '')

          return lang + extractedMeta
        }
      }

      return filename.split('.').pop() + meta
    }

    state.line = startLine + 1

    const token = state.push('fence', 'code', 0)
    token.info = langExtWithMeta()
    token.content = content
    token.markup = '```'
    token.map = [startLine, startLine + 1]

    return true
  }

  md.block.ruler.before('fence', 'snippet', parser)
}
