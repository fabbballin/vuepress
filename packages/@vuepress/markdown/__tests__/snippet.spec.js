import {
  Md,
  getFragment
} from './util'
import snippet from '../lib/snippet.js'
import highlightLines from '../lib/highlightLines.js'

const md = Md().use(snippet)
const mdH = Md()
  .use(highlightLines)
  .use(snippet)

describe('snippet', () => {
  test('import snippet', async () => {
    const input = await getFragment('code-snippet')
    const output = md.render(input)
    expect(output).toMatchSnapshot()
  })

  test('import snippet with highlight single line', async () => {
    const input = await getFragment('code-snippet-highlightLines-single')
    const output = mdH.render(input)
    expect(output).toMatchSnapshot()
  })

  test('import snippet with highlight multiple lines', async () => {
    const input = await getFragment('code-snippet-highlightLines-multiple')
    const output = mdH.render(input)
    expect(output).toMatchSnapshot()
  })

  test('import snippet with language specified as ruby', async () => {
    const input = await getFragment('code-snippet-specify-language-ruby')

    const output = md.render(input)
    expect(output).toMatchSnapshot()
    expect(output).toMatch(/ruby/)
  })

  test('import snippet with highlight lines and language specified as ruby', async () => {
    const input = await getFragment('code-snippet-highlightLines-single-specify-language-ruby')

    const output = mdH.render(input)
    expect(output).toMatchSnapshot()
  })
})
