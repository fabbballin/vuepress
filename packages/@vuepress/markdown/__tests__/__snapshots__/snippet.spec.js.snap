// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snippet import snippet 1`] = `
<pre><code class="language-js">export default function () {
  // ..
}
</code></pre>
`;

exports[`snippet import snippet with highlight lines and language specified as ruby 1`] = `
<div class="highlight-lines">
  <div class="highlighted">&nbsp;</div>
  <br>
  <div class="highlighted">&nbsp;</div>
  <br>
</div>def vuepress puts 'vuepress' end
`;

exports[`snippet import snippet with highlight multiple lines 1`] = `
<div class="highlight-lines">
  <div class="highlighted">&nbsp;</div>
  <div class="highlighted">&nbsp;</div>
  <div class="highlighted">&nbsp;</div>
  <br>
</div>export default function () { // .. }
`;

exports[`snippet import snippet with highlight single line 1`] = `
<div class="highlight-lines">
  <div class="highlighted">&nbsp;</div>
  <br>
  <div class="highlighted">&nbsp;</div>
  <br>
</div>export default function () { // .. }
`;

exports[`snippet import snippet with language specified as ruby 1`] = `
<pre><code class="language-ruby">def vuepress
  puts 'vuepress'
end
</code></pre>
`;
